### DISCUSSION TITLE
admin, sales report with graphs:  year display different on hosting/local

### DISCUSSION THREAD

**Discussion by torvista** [zencart/zencart#5822] [Ideas]
Hosting: php 8.1.4, Centos
![image](https://user-images.githubusercontent.com/4391026/212548884-b611cb0b-9084-427a-a047-5845fd08a8c2.png)
Local: php 8.1.4/8.2, Windows
![image](https://user-images.githubusercontent.com/4391026/212549004-29759c81-7f46-453b-ac21-9cb4565ce2cc.png)

Identical database.

I've dug down as far as here, from where the error originates:
https://github.com/zencart/zencart/blob/5fb07cf03e3d0e3d7817a8c3411d0e2345147d5d/includes/classes/zcDate.php#L175-L182

Echoing the $this->dateObject->getCalendarObject()
## Hosting/Bad:

> $this->dateObject->getCalendarObject() (object):
> IntlGregorianCalendar Object
> (
>     [valid] => 1
>     [type] => gregorian
>     [timeZone] => Array
>         (
>             [valid] => 1
>             [id] => Europe/Madrid
>             [rawOffset] => 3600000
>             [currentOffset] => 3600000
>         )
> 
>     [locale] => en
>     [fields] => Array
>         (
>             [era] => 1
>             [year] => 2023
>             [month] => 0
>             [week of year] => 2
>             [week of month] => 2
>             [day of year] => 15
>             [day of month] => 15
>             [day of week] => 1
>             [day of week in month] => 3
>             [AM/PM] => 1
>             [hour] => 4
>             [hour of day] => 16
>             [minute] => 21
>             [second] => 12
>             [millisecond] => 711
>             [zone offset] => 3600000
>             [DST offset] => 0
>             [year for week of year] => 2023
>             [localized day of week] => 7
>             [extended year] => 2023
>             [julian day] => 2459960
>             [milliseconds in day] => 58872711
>             [is leap month] => 0
>         )
> 
> )
> $output (string):
> Jan 22

## Local/Good:

> $this->dateObject->getCalendarObject() (object):
> IntlGregorianCalendar Object
> (
>     [valid] => 1
>     [type] => gregorian
>     [timeZone] => Array
>         (
>             [valid] => 1
>             [id] => Europe/Madrid
>             [rawOffset] => 3600000
>             [currentOffset] => 3600000
>         )
> 
>     [locale] => en_GB
>     [fields] => Array
>         (
>             [era] => 1
>             [year] => 2023
>             [month] => 0
>             [week of year] => 3
>             [week of month] => 3
>             [day of year] => 15
>             [day of month] => 15
>             [day of week] => 1
>             [day of week in month] => 3
>             [AM/PM] => 1
>             [hour] => 4
>             [hour of day] => 16
>             [minute] => 22
>             [second] => 29
>             [millisecond] => 287
>             [zone offset] => 3600000
>             [DST offset] => 0
>             [year for week of year] => 2023
>             [localized day of week] => 7
>             [extended year] => 2023
>             [julian day] => 2459960
>             [milliseconds in day] => 58949287
>             [is leap month] => 0
>         )
> 
> )
> 
> $output (string):
> Jan 23

Note that the only difference is the Local uses en_GB and Hosting en and the weeks of the year are different!
So I manually used en_GB instead of $this->locale and 
Hosting: locale changed as expected to en_GB, and weeks of year at 2 and year at Jan 22.
Local: although the locale was already/previously en_GB, the weeks of year changed to 2 and output to Jan 22: it should not have changed at all!

I'm not doing any more on this.

**Comment by piloujp** [comment]
Your date objects look good, it would be interesting to see the value of $converted_format. For GB date it should be like d/m/y h:mi:s . I am ready to bet it is OK.
Problem might be in stats_sales_report_graphs.php starting from line 112-113. The output date is used here and transmit to some Google code to build graphs. Substr() is involved too which is probably the source of the error if format is not as expected. Google code try to fix problems making completely unexpected results. In Japanese I had to modify substring values.
Difference between Your Windows version and Centos server is weird, probably due to the different Centos system date format.

**Comment by torvista** [comment]
The value of $converted_format is the same for both.
I suspected it was something to do with the locale...and when I change my local shop admin to spanish...it changes from "Jan 23" to "ene 22".
So, on the right track.

**Comment by piloujp** [comment]
Here is the code I modified in ...\admin\stats_sales_report_graphs.php after line 112, it won't work for you but might gives you an idea:
```
if ($_SESSION['language'] == 'japanese') {
  if ($sales_report_view == statsSalesReportGraph::YEARLY_VIEW && $report->size > 4) {
    echo substr($report->info[$i]['text'], 0, 4);
  } elseif ($sales_report_view == statsSalesReportGraph::MONTHLY_VIEW) {
    echo substr($report->info[$i]['text'], 0, 5);
  } elseif ($sales_report_view == statsSalesReportGraph::WEEKLY_VIEW) {
    echo substr($report->info[$i]['text'], 5, 5) . '～' . substr($report->info[$i]['text'], 18, 5);
  } elseif ($sales_report_view == statsSalesReportGraph::HOURLY_VIEW) {
    echo preg_replace('/0(?=[0-9])/','',substr($report->info[$i]['text'], 0, 2));
  } elseif ($report->size > 5) {
    echo substr($report->info[$i]['text'], 5, 5);
  } else {
    echo substr($report->info[$i]['text'], 5, 5);
  }
} else {
  if ($sales_report_view == statsSalesReportGraph::YEARLY_VIEW && $report->size > 5) {
    echo substr($report->info[$i]['text'], 0, 4);
  } elseif ($sales_report_view == statsSalesReportGraph::MONTHLY_VIEW) {
    echo substr($report->info[$i]['text'], 0, 3);
  } elseif ($sales_report_view == statsSalesReportGraph::WEEKLY_VIEW) {
    echo substr($report->info[$i]['text'], 0, 5) . '-' . substr($report->info[$i]['text'], 13, 5);
  } elseif ($sales_report_view == statsSalesReportGraph::HOURLY_VIEW) {
    echo preg_replace('/0(?=[0-9])/','',substr($report->info[$i]['text'], 0, 2));
  } elseif ($report->size > 5) {
    echo substr($report->info[$i]['text'], 0, 5);
  } else {
    echo substr($report->info[$i]['text'], 0, 5);
  }
}
```
For British local you can't test on language but important part is in substring function values. The code above works for Japanese and English US.  Locale and date format settings in languages files are important too of course, the code above completely depends on it.

**Comment by piloujp** [comment]
I think you can pass my precedent post. I realized the substring stuff is only for cosmetic display and can be removed in many case as $report->info[] returns already formated parts of dates.
Your problem comes from locale setting somewhere. You could check if you don't use a 'o' parameter instead of 'Y' for Date() function somewhere. Although it should only has influence on first week of January. Same with 'Y' instead of 'y' for your date object. Oooh, I just remembered that the ZCdate.php classe had a mistake. Line 114 and 115:
```
                '%y' => 'YY',
                '%Y' => 'YYYY',
```
Should be :
```
                '%y' => 'yy',
                '%Y' => 'yyyy', // or 'y', it is identical
```
Try modifying this. y is standard and Y is for year in “week of year”.

**Comment by torvista** [comment]
Thanks for taking the time to respond/your interest.

Consequently I've been looking at it and gradually homing in on the cause, but yes **this has fixed it**:

                '%y' => 'YY',
                '%Y' => 'YYYY',
Should be :

                '%y' => 'yy',
                '%Y' => 'yyyy', // or 'y', it is identical

Please PR!!

**Comment by mc12345678** [comment]
Another factor of consideration here where a *nux system is compared to a windows system is the potential that the windows system supports database data return of integers and floats may truly come back as integers and floats instead of text. 

May I suggest going to the admin side catalog and observing if all of the product and categories appear to be disabled, though on the catalog side at least some of those are active. It would at least eliminate a potential issue caused by some sort of strict comparison somewhere in the associated code.

**Comment by torvista** [comment]
Eh? the last few comments have disappeared...including the fix for this issue!

**Comment by piloujp** [comment]
Please just pass on precedent post. I realized it is just for cosmetic display. Problem is from some date format setting somewhere using year in “week of year” instead of standard year. I think I found it but you have to test that. In classe zcDate.php replace lines 114-115 :
```
                '%y' => 'YY',
                '%Y' => 'YYYY',
```
by :
```
                '%y' => 'yy',
                '%Y' => 'yyyy',
```
or :
```
                '%y' => 'yy',
                '%Y' => 'y',
```
The upper case Y in intldate object is for year in “week of year”. It should be lowercase! I did not have this problem because I use my modified zcDate classe...

**Comment by piloujp** [comment]
Strange when I came back to see if you posted something It had disappeared but I wrote it and sent it again. I will PR this one with success I hope but after some sleep.
I agree about strict comparaison that cause trouble too often but I checked catalog on admin side and everything is OK. Actually if I understand what you wrote it should be checked on a Centos no?

**Comment by torvista** [comment]
Nevermind, I've done it.

**Comment by proseLA** [comment]
i'm still not convinced here.  this should be an "easy" test.  here is the code:

```
    $timestamp = time();
    $locale = 'en';
    $format = 'Y';
    $fmt = new IntlDateFormatter(
        $locale,
        IntlDateFormatter::FULL,
        IntlDateFormatter::FULL,
        date_default_timezone_get(),
        IntlDateFormatter::GREGORIAN,
        $format
    );
    echo "-------->" . $fmt->format($timestamp) . "<---------<br>";
```

now change the locale and the format so that you are getting '22' for year.  and report back.

and if you can reproduce that, then give us the results of:

```
echo "-------->" . date_default_timezone_get() . "<---------<br>";
```

i'm not saying you are wrong, but i am not able to reproduce it,  i find it hard to imagine a system reporting back `2022` or `22` for a year based on a date of today with an output format of a capital `Y`; or any combination of capital `Y`s.

you can see that [the intl date format uses iso date format codes](https://framework.zend.com/manual/1.12/en/zend.date.constants.html#zend.date.constants.selfdefinedformats), and capital `Y`s are more than valid.

**Comment by torvista** [comment]
You are right to query this.
I expanded your test code....and stuck it just above the incorrect output on the sales report with graphs

LOCAL HOSTING with admin in locale es-ES (windows) and the incorrect output below. Admin in **english** is ok.
 
![2023-01-18 12_35_13-Admin Stats Sales Report Graphs — Mozilla Firefox](https://user-images.githubusercontent.com/4391026/213162338-2d37cb5a-6d51-4cdc-9b4c-86eb966a6bb6.gif)

This is part of the $report object
![image](https://user-images.githubusercontent.com/4391026/213162634-216035bb-00a7-4204-9847-16c26287eaa5.png)


REMOTE HOSTING with admin in locale en_GB.utf8 and the incorrect output below. Admin in **spanish** is ok.

![2023-01-18 12_42_28-Admin Stats Sales Report Graphs](https://user-images.githubusercontent.com/4391026/213162918-d46d6a1e-0e46-42e7-98a0-fd59a5c15e21.gif)

![image](https://user-images.githubusercontent.com/4391026/213163058-f3b199e9-9240-4b21-bc32-cc664f3e4545.png)

hours of fun for all the family....

**Comment by torvista** [comment]
Here is the zcDate debug. I note the locale is always "C", and at the end, the misbehaving output "ene 22" has a different timestamp to the rest.

> [18-Jan-2023 16:48:59 Europe/Madrid] zcDate construction: 
> \zcDate::__set_state(array(
>    'observerAliases' => 
>   array (
>     'NOTIFIY_ORDER_CART_SUBTOTAL_CALCULATE' => 'NOTIFY_ORDER_CART_SUBTOTAL_CALCULATE',
>   ),
>    'useIntlDate' => true,
>    'useStrftime' => false,
>    'locale' => 'C',
>    'strftime2date' => 
>   array (
>     'from' => 
>     array (
>       0 => '%a',
>       1 => '%A',
>       2 => '%b',
>       3 => '%B',
>       4 => '%d',
>       5 => '%H',
>       6 => '%m',
>       7 => '%M',
>       8 => '%S',
>       9 => '%T',
>       10 => '%x',
>       11 => '%X',
>       12 => '%y',
>       13 => '%Y',
>     ),
>     'to' => 
>     array (
>       0 => 'D',
>       1 => 'l',
>       2 => 'M',
>       3 => 'F',
>       4 => 'd',
>       5 => 'H',
>       6 => 'm',
>       7 => 'i',
>       8 => 's',
>       9 => 'H:i:s',
>       10 => 'm/d/Y',
>       11 => 'H:i:s',
>       12 => 'y',
>       13 => 'Y',
>     ),
>   ),
>    'strftime2intl' => 
>   array (
>     'from' => 
>     array (
>       0 => '%a',
>       1 => '%A',
>       2 => '%b',
>       3 => '%B',
>       4 => '%d',
>       5 => '%H',
>       6 => '%m',
>       7 => '%M',
>       8 => '%S',
>       9 => '%T',
>       10 => '%x',
>       11 => '%X',
>       12 => '%y',
>       13 => '%Y',
>     ),
>     'to' => 
>     array (
>       0 => 'E',
>       1 => 'EEEE',
>       2 => 'MMM',
>       3 => 'MMMM',
>       4 => 'dd',
>       5 => 'HH',
>       6 => 'MM',
>       7 => 'mm',
>       8 => 'ss',
>       9 => 'HH:mm:ss',
>       10 => 'dd/MM/y',
>       11 => 'HH:mm',
>       12 => 'YY',
>       13 => 'YYYY',
>     ),
>   ),
>    'debug' => true,
>    'dateObject' => NULL,
> ))
> 
> [18-Jan-2023 16:49:00 Europe/Madrid] zcDate output for '%d/%m/%Y %H:%M:%S' with timestamp (1674056940), with format converted to 'dd/MM/YYYY HH:mm:ss': '"18\/01\/2023 16:49:00"'
> 
> [18-Jan-2023 16:49:00 Europe/Madrid] zcDate output for '%d/%m/%Y %H:%M:%S' with timestamp (1674056940), with format converted to 'dd/MM/YYYY HH:mm:ss': '"18\/01\/2023 16:49:00"'
> 
> [18-Jan-2023 16:49:00 Europe/Madrid] zcDate output for '%b %y' with timestamp (1672527600), with format converted to 'MMM YY': '"ene 22"'
> 
> [18-Jan-2023 16:49:00 Europe/Madrid] zcDate output for '%A %d %b %Y %H:%M' with timestamp (1674056940), with format converted to 'EEEE dd MMM YYYY HH:mm': '"mi\u00e9rcoles 18 ene 2023 16:49"'

**Comment by proseLA** [comment]
[18-Jan-2023 16:49:00 Europe/Madrid] zcDate output for '%b %y' with timestamp (1672527600), with format converted to 'MMM YY': '"ene 22"'

is the problem no?

being able to replicate this can allow us to fix it.  **but until we can replicate this timestamp and output, there will be no fix!**

**Comment by torvista** [comment]
and going to the previous page, which is the monthly for the previous year:
![image](https://user-images.githubusercontent.com/4391026/213222372-1e6287be-3c6d-479e-a786-8c4972ab5464.png)

**Comment by proseLA** [comment]
that is the problem timestamp and format.  the question is what is the locale set to.

**Comment by proseLA** [comment]
> and going to the previous page, which is the monthly for the previous year: ![image](https://user-images.githubusercontent.com/4391026/213222372-1e6287be-3c6d-479e-a786-8c4972ab5464.png)

get rid of this and please stop posting images.

**Comment by torvista** [comment]
'"ene 22"' is the problem 
locale is es-ES (windows)

What's the problem with images?

**Comment by proseLA** [comment]
do you want help or not?  images suck here.  can not copy timestamps of which i had already narrowed it down to the one i quoted above.

i'm trying to narrow down a problem and you are now trying to widen it out.

this is a windows server that you are having problems with?

**Comment by torvista** [comment]
> do you want help or not?

To be honest I don't care. I'm reporting a problem and providing as much information as I can.
I'm not forced to report it, and no-one is forced to fix it.
I'm sure we both have more important things to do, as you've reminded me.

**Comment by piloujp** [comment]
I hope everybody cooled down. I have the solution and I am going to explain it.
Y parameter instead of y is the main problem as already stated but the system time zone/locale is although a problem if different from Zen Cart settings.
First some important data you need to know:
- en_US and en_GB locales have different first day of week. American locale starts weeks on Sunday and Great Britain one starts on Monday.
- When using Y parameter new year begins on a 'first day of the week' which is most of the time not  the first of January.
- With y parameter, year always begins on the first of January whatever the day of the week.

What made the problem more difficult to identify is that this year (2023) January the first is on Sunday. In the us_EN locale Sunday is although the first day of the week which means whatever parameter you use, y or Y, results are identical.
The statistic display graph by month most certainly uses the first day of the month to get a timestamp, not the actual date.
When you use a locale like en_GB, es_ES and many others that start weeks on Monday there is one day difference between y and Y parameters for the start of the year 2023. With Y parameter it starts on Monday the 2nd and with y parameter on Sunday the 1st. With Y parameter Sunday is still in 2022 which is why you get 'Jan 22'.

Finally to get a time stamp mktime function is generally used but it is dependent on PHP default time zone, not what is set in date object. Getting a time stamp at 0h might not be 0h at all. For example, I am at GMT +9, If I set Date object time zone to 'Europe/Madrid' and use `mktime(0, 0, 0, 1, 1, 2023)` I will get a timestamp for Dec 31 2022 16h. I have to use `mktime(8, 0, 0, 1, 1, 2023)` to get right result. 8 is the difference between GMT +9 (Tokyo) and GMT +1 (Madrid). This, mixed with Y parameter, can makes things very hard to debug.

I made a small PHP script for illustration and testing on your system. Looking at Internet forums, it seems it is sometimes difficult to set locale and time zone on Centos systems.

Please try this :
```
<?php
// date_default_timezone_set('Europe/Madrid'); // For testing when in a different timezone
echo "Your system (PHP) time zone is " . date_default_timezone_get() . "\n";
$DateObject = datefmt_create(
    'en_GB',
    IntlDateFormatter::FULL,
    IntlDateFormatter::FULL,
    'Europe/Madrid',
    IntlDateFormatter::GREGORIAN,
);
$DateObject->setPattern("MMM yy '       Day of week:' e eee");
echo "Locale = en_GB    Patern = yy :\n";
$testdate = mktime(-1, 0, 0, 1, 1, 2023);
echo '2023/01/01 -1h : ' . $DateObject->format($testdate) . "\n";
$testdate = mktime(0, 0, 0, 1, 1, 2023);
echo '2023/01/01 0h : ' . $DateObject->format($testdate) . "\n";

$DateObject->setPattern("MMM YY '       Day of the week:' e eee");
echo "Locale = en_GB    Patern = YY :\n";
$testdate = mktime(-1, 0, 0, 1, 1, 2023);
echo '2023/01/01 -1h : ' . $DateObject->format($testdate) . "\n";
$testdate = mktime(0, 0, 0, 1, 1, 2023);
echo '2023/01/01 0h : ' . $DateObject->format($testdate) . "\n";
$testdate = mktime(0, 0, 0, 1, 2, 2023);
echo '2023/01/02 0h : ' . $DateObject->format($testdate) . "\n";

// New locale
$DateObject = datefmt_create(
    'en_US',
    IntlDateFormatter::FULL,
    IntlDateFormatter::FULL,
    'Europe/Madrid',
    IntlDateFormatter::GREGORIAN,
);
$DateObject->setPattern("MMM yy '       Day of week:' e eee");
echo "\nLocale = en_US    Patern = yy :\n";
$testdate = mktime(-1, 0, 0, 1, 1, 2023);
echo '2023/01/01 -1h : ' . $DateObject->format($testdate) . "\n";
$testdate = mktime(0, 0, 0, 1, 1, 2023);
echo '2023/01/01 0h : ' . $DateObject->format($testdate) . "\n";

$DateObject->setPattern("MMM YY '       Day of the week:' e eee");
echo "Locale = en_US    Patern = YY :\n";
$testdate = mktime(-1, 0, 0, 1, 1, 2023);
echo '2023/01/01 -1h : ' . $DateObject->format($testdate) . "\n";
$testdate = mktime(0, 0, 0, 1, 1, 2023);
echo '2023/01/01 0h : ' . $DateObject->format($testdate) . "\n";
$testdate = mktime(0, 0, 0, 1, 2, 2023);
echo '2023/01/02 0h : ' . $DateObject->format($testdate) . "\n";

?>
```

**Comment by torvista** [comment]
Very enlightening!
I added more locales
# LOCAL/Windows

Your system (PHP) time zone is Europe/Madrid

Locale = en_US Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 7 Sat
2023/01/01 0h : Jan 23 Day of week: 1 Sun

Locale = en_US Pattern = YY :
2023/01/01 -1h : Dec 22 Day of the week: 7 Sat
2023/01/01 0h : Jan 23 Day of the week: 1 Sun
2023/01/02 0h : Jan 23 Day of the week: 2 Mon

Locale = en_US.utf8 Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 7 Sat
2023/01/01 0h : Jan 23 Day of week: 1 Sun

Locale = en_US.utf8 Pattern = YY :
2023/01/01 -1h : Dec 22 Day of the week: 7 Sat
2023/01/01 0h : Jan 23 Day of the week: 1 Sun
2023/01/02 0h : Jan 23 Day of the week: 2 Mon

Locale = en Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 7 Sat
2023/01/01 0h : Jan 23 Day of week: 1 Sun

Locale = en Pattern = YY :
2023/01/01 -1h : Dec 22 Day of the week: 7 Sat
2023/01/01 0h : Jan 23 Day of the week: 1 Sun
2023/01/02 0h : Jan 23 Day of the week: 2 Mon

Locale = English_United States.1252 Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 6 Sat
2023/01/01 0h : Jan 23 Day of week: 7 Sun

Locale = English_United States.1252 Pattern = YY :
2023/01/01 -1h : Dec 23 Day of the week: 6 Sat
2023/01/01 0h : Jan 23 Day of the week: 7 Sun
2023/01/02 0h : Jan 23 Day of the week: 1 Mon

Locale = en_GB.utf8 Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 6 Sat
2023/01/01 0h : Jan 23 Day of week: 7 Sun

Locale = en_GB.utf8 Pattern = YY :
2023/01/01 -1h : Dec 22 Day of the week: 6 Sat
2023/01/01 0h : Jan 22 Day of the week: 7 Sun
2023/01/02 0h : Jan 23 Day of the week: 1 Mon

Locale = es_ES.UTF8 Pattern = yy :
2023/01/01 -1h : dic 22 Day of week: 6 sáb
2023/01/01 0h : ene 23 Day of week: 7 dom

Locale = es_ES.UTF8 Pattern = YY :
2023/01/01 -1h : dic 22 Day of the week: 6 sáb
2023/01/01 0h : ene 22 Day of the week: 7 dom
2023/01/02 0h : ene 23 Day of the week: 1 lun

Locale = es-ES Pattern = yy :
2023/01/01 -1h : dic 22 Day of week: 6 sáb
2023/01/01 0h : ene 23 Day of week: 7 dom

Locale = es-ES Pattern = YY :
2023/01/01 -1h : dic 22 Day of the week: 6 sáb
2023/01/01 0h : ene 22 Day of the week: 7 dom
2023/01/02 0h : ene 23 Day of the week: 1 lun

Locale = Spanish_Spain.1252 Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 6 Sat
2023/01/01 0h : Jan 23 Day of week: 7 Sun

Locale = Spanish_Spain.1252 Pattern = YY :
2023/01/01 -1h : Dec 23 Day of the week: 6 Sat
2023/01/01 0h : Jan 23 Day of the week: 7 Sun
2023/01/02 0h : Jan 23 Day of the week: 1 Mon

Locale = es Pattern = yy :
2023/01/01 -1h : dic 22 Day of week: 6 sáb
2023/01/01 0h : ene 23 Day of week: 7 dom

Locale = es Pattern = YY :
2023/01/01 -1h : dic 22 Day of the week: 6 sáb
2023/01/01 0h : ene 22 Day of the week: 7 dom
2023/01/02 0h : ene 23 Day of the week: 1 lun

# REMOTE/Centos
Your system (PHP) time zone is Europe/Madrid

Locale = en_US Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 7 Sat
2023/01/01 0h : Jan 23 Day of week: 1 Sun

Locale = en_US Pattern = YY :
2023/01/01 -1h : Dec 22 Day of the week: 7 Sat
2023/01/01 0h : Jan 23 Day of the week: 1 Sun
2023/01/02 0h : Jan 23 Day of the week: 2 Mon

Locale = en_US.utf8 Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 7 Sat
2023/01/01 0h : Jan 23 Day of week: 1 Sun

Locale = en_US.utf8 Pattern = YY :
2023/01/01 -1h : Dec 22 Day of the week: 7 Sat
2023/01/01 0h : Jan 23 Day of the week: 1 Sun
2023/01/02 0h : Jan 23 Day of the week: 2 Mon

Locale = en Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 7 Sat
2023/01/01 0h : Jan 23 Day of week: 1 Sun

Locale = en Pattern = YY :
2023/01/01 -1h : Dec 22 Day of the week: 7 Sat
2023/01/01 0h : Jan 23 Day of the week: 1 Sun
2023/01/02 0h : Jan 23 Day of the week: 2 Mon

Locale = English_United States.1252 Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 6 Sat
2023/01/01 0h : Jan 23 Day of week: 7 Sun

Locale = English_United States.1252 Pattern = YY :
2023/01/01 -1h : Dec 23 Day of the week: 6 Sat
2023/01/01 0h : Jan 23 Day of the week: 7 Sun
2023/01/02 0h : Jan 23 Day of the week: 1 Mon

Locale = en_GB.utf8 Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 6 Sat
2023/01/01 0h : Jan 23 Day of week: 7 Sun

Locale = en_GB.utf8 Pattern = YY :
2023/01/01 -1h : Dec 22 Day of the week: 6 Sat
2023/01/01 0h : Jan 22 Day of the week: 7 Sun
2023/01/02 0h : Jan 23 Day of the week: 1 Mon

Locale = es_ES.UTF8 Pattern = yy :
2023/01/01 -1h : dic 22 Day of week: 6 sáb
2023/01/01 0h : ene 23 Day of week: 7 dom

Locale = es_ES.UTF8 Pattern = YY :
2023/01/01 -1h : dic 22 Day of the week: 6 sáb
2023/01/01 0h : ene 22 Day of the week: 7 dom
2023/01/02 0h : ene 23 Day of the week: 1 lun

Locale = es-ES Pattern = yy :
2023/01/01 -1h : dic 22 Day of week: 6 sáb
2023/01/01 0h : ene 23 Day of week: 7 dom

Locale = es-ES Pattern = YY :
2023/01/01 -1h : dic 22 Day of the week: 6 sáb
2023/01/01 0h : ene 22 Day of the week: 7 dom
2023/01/02 0h : ene 23 Day of the week: 1 lun

Locale = Spanish_Spain.1252 Pattern = yy :
2023/01/01 -1h : Dec 22 Day of week: 6 Sat
2023/01/01 0h : Jan 23 Day of week: 7 Sun

Locale = Spanish_Spain.1252 Pattern = YY :
2023/01/01 -1h : Dec 23 Day of the week: 6 Sat
2023/01/01 0h : Jan 23 Day of the week: 7 Sun
2023/01/02 0h : Jan 23 Day of the week: 1 Mon

Locale = es Pattern = yy :
2023/01/01 -1h : dic 22 Day of week: 6 sáb
2023/01/01 0h : ene 23 Day of week: 7 dom

Locale = es Pattern = YY :
2023/01/01 -1h : dic 22 Day of the week: 6 sáb
2023/01/01 0h : ene 22 Day of the week: 7 dom
2023/01/02 0h : ene 23 Day of the week: 1 lun

Note that I do not advocate any effort to be made to account for Windows server behaviour, that is of academic interest only.

**Comment by piloujp** [comment]
Except something.1252 locales everything is as it should be. On my Win 10 system, something.1252 is not recognized and locale revert to system default (Japanese for me). I am not sure these are valid locales.
I think it is now near 100% sure that Parameter in zcDate need to be changed from YY to yy and YYYY to y (or yyyy).

**Comment by mc12345678** [comment]
I don't necessarily see it (especially based on the provided data) as being a windows, centos, or *nux specific issue. The link previously provided: https://framework.zend.com/manual/1.12/en/zend.date.constants.html#zend.date.constants.selfdefinedformats discusses this issue of using `y` versus `Y` which piloujp further described.

**Comment by piloujp** [comment]
> https://framework.zend.com/manual/1.12/en/zend.date.constants.html#zend.date.constants.selfdefinedformats discusses this issue of using `y` versus `Y` which piloujp further described.

Just to have the same reference and avoid confusion, the link above are ref. for old date object I think, not sure, but this is not what is used by datefmt_format(). Intl Date object uses similar parameters but different. Info on this hard to find in PHP manual but here is the link for IntlDateFormatter::format (alias datefmt_format()) format string parameters:
[https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table](https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table)

**Comment by piloujp** [comment]
Just to add precision on when 'bug' might appear here are information on 52-53 week year system.
53 weeks years are defined like this : In calendars with Monday as first day of the week, first week of the year always contains 4th of January then you count the number of week until next year first week (2015-2020-2026-2032). When weeks start on Sunday first week of the year always includes 1st of January (2011-2016-2022-2028-2033).
Making it short it means every five or six years (variation due to leap years) there is a 53 weeks year. It although means there are six days (dates) that will be end or beginning of the year depending on the year when using weeks to count years.
Week starting on Monday (GB, European countries, ...):
Dec. 29 to Jan. 3
Week starts on Sunday (US, Japan, ...):
Dec. 26 to Dec. 31
During these periods using week years will give wrong year on dates. Most of the 'bugs' will be display problem but it could be worst with Coupon creation ( end date before start date) or payment module like PayPal...
Of Course it can't be reproduced now except perhaps by tweaking/hacking your system date. I hope nobody is going to ask to wait until end of the year. ;)

**Comment by piloujp** [comment]
For simulation of concerned dates use this script in your PHP command line (not in browser). It is an updated version of the one posted above. On Windows systems this dates will be displayed with different color. Choose the year by modifying variable at beginning of the script. You can loop through 12 years if you uncomment the for loop at the beginning and end of file.
```
<?php
$year = 2023; // Set year to test or start testing with loop

//for ($k=0;$k<12;$k++) { // Automatic cycle trough years (uncomment here and at end of file)

// Set locales here, locale 1 with first day of the week on Monday and locale 2 with first day of the week on Sunday.
$locale1 = 'en_GB'; // weeks start on Monday

$locale2 = 'en_US'; // weeks start on Sunday

$zone = date_default_timezone_get();

echo "Your system (PHP) time zone is " . date_default_timezone_get() . "\n";
for ($j = 0; $j < 2; $j++) { // loop dor two locales
    $locale =  ($j == 0) ? $locale1 : $locale2;
    $DateObject = datefmt_create(
        $locale,
        IntlDateFormatter::FULL,
        IntlDateFormatter::FULL,
        $zone,
        IntlDateFormatter::GREGORIAN,
    );
    $DateObject->setPattern("MMM y '       D.of.W:' e eee");
    echo "Locale = " . datefmt_get_locale($DateObject) . "    Patern = y :\n";
    $testdate = mktime(0, 0, 0, 12, 31, $year-1);
    echo $year-1 . '/12/31 0h : ' . $DateObject->format($testdate) . "\n";
    $testdate = mktime(0, 0, 0, 1, 1, $year);
    echo "\033[32m" . $year . '/01/01 0h : ' . $DateObject->format($testdate) . "\033[0m\n";

    $DateObject->setPattern("MMM Y '       D.of.W:' e eee'    W.nb: 'w");
    echo "Locale = " . datefmt_get_locale($DateObject) . "    Patern = Y :\n";
    $mois = 12;
    $jour =  ($j == 0) ? 28 : 25;
    for ($i = 0; $i < 8; $i++) {
        If ($jour >20 && $jour <= 31) {
            $testdate = mktime(0, 0, 0, $mois, $jour, $year-1);
            $dateform = $DateObject->format($testdate);
            $DateObject->setPattern("Y");
            if ($DateObject->format($testdate) == $year) {
                echo "\033[31m" . $year-1 . '/' . sprintf("%'.02d",$mois) . '/' . sprintf("%'.02d",$jour) . ' 0h : ' . $dateform . "\033[0m\n";
            } else {
                echo $year-1 . '/' . sprintf("%'.02d",$mois) . '/' . sprintf("%'.02d",$jour) . ' 0h : ' . $dateform . "\n";
            }
            $DateObject->setPattern("MMM Y '       D.of.W:' e eee'    W.nb: 'w");
        } else {
            $testdate = mktime(0, 0, 0, $mois, $jour, $year);
            If ($jour == 1) {
                $dateform = $DateObject->format($testdate);
                $DateObject->setPattern("Y");
                if ($DateObject->format($testdate) == $year - 1) {
                    echo "\033[35m" . $year . '/' . sprintf("%'.02d",$mois) . '/' . sprintf("%'.02d",$jour) . ' 0h : ' . $dateform . "\033[0m\n";
                } else {
                    echo "\033[32m" . $year . '/' . sprintf("%'.02d",$mois) . '/' . sprintf("%'.02d",$jour) . ' 0h : ' . $dateform . "\033[0m\n";  
                }
                $DateObject->setPattern("MMM Y '       D.of.W:' e eee'    W.nb: 'w");
            } else {
                $dateform = $DateObject->format($testdate);
                $DateObject->setPattern("Y");
                if ($DateObject->format($testdate) == $year - 1) {
                    echo "\033[31m" . $year . '/' . sprintf("%'.02d",$mois) . '/' . sprintf("%'.02d",$jour) . ' 0h : ' . $dateform . "\033[0m\n";
                } else {
                    echo $year . '/' . sprintf("%'.02d",$mois) . '/' . sprintf("%'.02d",$jour) . ' 0h : ' . $dateform . "\n";
                }
                $DateObject->setPattern("MMM Y '       D.of.W:' e eee'    W.nb: 'w");
            }
        }
        if ($jour == 31) {
            $jour = 0;
            $mois = 1;
        }
        $jour++;
    }
    echo "\n";
}
// uncomment following for loop through years
/*$year++;
sleep(1);  // delay between dispay of each year
//usleep(500000);
}*/
?>
```
